### DISCUSSION TITLE
Customizable mirror device to allow for a hidden input or output

### DISCUSSION THREAD

**Discussion by TheRico12** [ExistentialAudio/BlackHole#619] [Q&A]
Hi! 
What do you mean with mirror device and hidden device? How can I customise them?

Thanks in advance for the support

**Comment by devinroth** [comment]
I haven't spent any time documenting it yet so here's the quick rundown.  Modify the following pre-compiler constants. 
```
// Original Device
kDevice_IsHidden
kDevice_HasInput
kDevice_HasOutput

// Mirrored Device
kDevice2_IsHidden
kDevice2_HasInput
kDevice2_HasOutput
```

When all are set to ```true``` a 2nd BlackHole will show up that works exactly the same. The inputs and outputs are mirrored so the outputs from both devices go to the inputs of both devices. 

This is useful if you need a separate device for input and output. 

Example 

```
// Original Device
kDevice_IsHidden=false
kDevice_HasInput=true
kDevice_HasOutput= false

// Mirrored Device
kDevice2_IsHidden=false
kDevice2_HasInput= false
kDevice2_HasOutput=true
```

In this situation we have two BlackHole devices. One will have inputs only and the other will have outputs only. 

The way I usually use this in projects is hide the mirrored device and use it behind the scenes. That way the user will see an input only device while I route audio through to the output behind them scenes. 

Hope that makes sense.

**Reply by TheRico12** [reply]
@devinroth thank you so much for the quick reply and explanation! This is exactly what I was searching for (having separate I/O).
I think I've understood the logic behind, but it's not clear to me how can I use the solution that you suggested me with the mirrored one hidden...if I do that, I can only see one BH instance with inputs only, so I don't get how can I route the the outputs instead. I hope that my bad English is understandable. 

Thanks again

**Reply by devinroth** [reply]
To access a hidden device you need to request the AudioDeviceID using `kAudioHardwarePropertyTranslateUIDToDevice`. Once you have the AudioDeviceID you can use it normally with any of the Apple Audio APIs. The problem is many of the 3rd party audio APIs (JUCE, Electron, etc) don't support hidden devices. Although you can probably modify those APIs to make them work. They all rely on Apple's CoreAudio at some level.

**Reply by devinroth** [reply]
I updated the README to include more details on how this can be implemented. https://github.com/ExistentialAudio/BlackHole#developer-guides

**Reply by TheRico12** [reply]
Thank you so much,  but this is too much for my knowledge! It's like chinese to me ;)
